lib/scholar/neighbors/brute_knn.ex

```diff
@@ -0,0 +1,230 @@
+defmodule Scholar.Neighbors.BruteKNN do
+  @moduledoc """
+  Brute-Force k-Nearest Neighbor Search Algorithm.
+
+  In order to find the k-nearest neighbors the algorithm calculates
+  the distance between the query point and each of the data samples.
+  Therefore, its time complexity is $O(MN)$ for $N$ samples and $M$ query points.
+  It uses $O(BN)$ memory for batch size $B$.
+  Larger batch sizes will lead to faster predictions, but will consume more memory.
+  """
+  import Nx.Defn
+  import Scholar.Shared
+  require Nx
+
+  @derive {Nx.Container, keep: [:num_neighbors, :metric, :batch_size], containers: [:data]}
+  defstruct [:num_neighbors, :metric, :data, :batch_size]
+
+  opts = [
+    num_neighbors: [
+      required: true,
+      type: :pos_integer,
+      doc: "The number of nearest neighbors."
+    ],
+    metric: [
+      type: {:or, [{:custom, Scholar.Options, :metric, []}, {:fun, 2}]},
```

Comment:
Perhaps `Scholar.Options.metric` should be edited to support functions of arity 2?

Reply:
If everywhere we accept a metric we also accept functions, then yes. The easiest is probably to make it so it always returns a 2-arity function and then we call it. :)